src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatabilityAnalyzer.cs

```diff
@@ -220,26 +220,34 @@ static ImmutableArray<IMethodSymbol> GetOperatingSystemGuardMethods(IMethodSymbo
             });
         }
 
-        private static bool HasInterproceduralResult(IOperation platformSpecificOperation, SmallDictionary<string, PlatformAttributes> attributes,
+        private static bool HasGuardedLambdaOrLocalFunctionResult(IOperation platformSpecificOperation, SmallDictionary<string, PlatformAttributes> attributes,
             DataFlowAnalysisResult<GlobalFlowStateBlockAnalysisResult, GlobalFlowStateAnalysisValueSet> analysisResult)
         {
-            if (platformSpecificOperation.IsWithinLambdaOrLocalFunction())
+            if (!platformSpecificOperation.IsWithinLambdaOrLocalFunction(out var containingLambdaOrLocalFunctionOperation))
             {
-                var results = analysisResult.TryGetInterproceduralResults();
-                if (results != null)
+                return false;
+            }
+
+            var results = analysisResult.TryGetLambdaOrLocalFunctionResults(containingLambdaOrLocalFunctionOperation);
+            Debug.Assert(results.Any(), "Expected at least one analysis result for lambda/local function");
+
+            foreach (var localResult in results)
+            {
+                Debug.Assert(localResult.ControlFlowGraph.OriginalOperation == containingLambdaOrLocalFunctionOperation);
+
+                var localValue = localResult[platformSpecificOperation.Kind, platformSpecificOperation.Syntax];
+
+                // Value must be known and guarded in all analysis contexts.
+                // NOTE: IsKnownValueGuarded mutates the input values, so we pass in cloned values
```

Comment:
@buyaa-n I think we should prioritize this cleanup item to remove side effects from `IsKnownValueGuarded` and make `PlatformAttributes` type immutable. This will very likely lead to more subtle bugs, like the one I fixed in the latest commit.

Reply:
Agree with making immutable entirely, but for the current context updating the attributes for the current context was intentional. Like the guard method might only suppress the part of attributes and we might still want to warn for the not suppressed attribute

Reply:
The issue with that is if there are 2 calls to a local function, where first one is guarded but second is not, then we do not warn. That is fixed now though.

Reply:
I see, as it is returning false when the value is not guarded (`!IsKnownValueGuarded`) i think it still can fail in case the order of the results of local functions are switched, no?

Reply:
I'll add a test, but I think given all calls to `IsKnownValueGuarded` now operate on a copy it shouldn't matter. Mutation does make this harder to reason about, so the best way to verify would be through a unit test.

Reply:
As it is only for looping through lambda/local function results I am good with working with a copy

Reply:
@buyaa-n added test to verify it works for either order now.

Reply:
Thanks